### PR TITLE
PP-4769 Telephone Payment Response

### PR DESCRIPTION
**Pull request by nimalank7**
## WHAT YOU DID

This PR makes the payment_outcome status field mandatory. The status field is either "Success" or "Failed" any other value is rejected. if the status is "Success" then no other fields are required. If status is failed then code is required with the requirement that the code must be one of "P0010", "P0030" and "P0050". The response reflects these values back. I have created a CreateTelephonePaymentService which does this.

The supplemental section is optional in all cases. The payment_id returned for a successful payment is currently "DUMMY API - RECORD NOT PERSISTED".

These features are enforced using custom validation over several fields (in PaymentOutcomeValidator). There is a resource test for testing this validation.

I have modified the GetTelephonePaymentIT to test for the resulting JSON instead of merely checking for a 201 response. I have also subsequently removed the representation tests as they are covered by the integration tests.

### CODE REVIEW

File: src/main/java/uk/gov/pay/api/model/telephone/State.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package uk.gov.pay.api.model.telephone;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class State {
+
+    @JsonProperty("status")
```
Comment (mrlumbu) [inline src/main/java/uk/gov/pay/api/model/telephone/State.java]: you don't need these annotations if there is no difference between field name and json property name

File: src/main/java/uk/gov/pay/api/model/telephone/TelephonePaymentResponse.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
+package uk.gov.pay.api.model.telephone;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+public class TelephonePaymentResponse {
```
Comment (mrlumbu) [inline src/main/java/uk/gov/pay/api/model/telephone/TelephonePaymentResponse.java]: you could remove all the `JsonProperty` annotations and just annotate the class with `@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)`

File: src/main/java/uk/gov/pay/api/model/telephone/TelephonePaymentResponse.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    private String paymentId;
+
+    @JsonProperty("state")
+    private State state;
+
+    public TelephonePaymentResponse() {
+        // For Jackson
+    }
+
+    public TelephonePaymentResponse(int amount, String reference, String description, String createdDate, String authorisedDate, String processorId, String providerId, String authCode, PaymentOutcome paymentOutcome, String cardType, String nameOnCard, String emailAddress, String cardExpiry, String lastFourDigits, String firstSixDigits, String telephoneNumber, String paymentId, State state) {
```
Comment (mrlumbu) [inline src/main/java/uk/gov/pay/api/model/telephone/TelephonePaymentResponse.java]: maybe break over multiple lines?

File: src/main/java/uk/gov/pay/api/service/telephone/CreateTelephonePaymentService.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                        s.getErrorCode(),
+                        s.getErrorMessage()
+                )).orElse(null);
+        
+        PaymentOutcome paymentOutcome = new PaymentOutcome(
+                createTelephonePaymentRequest.getPaymentOutcome().getStatus(),
+                createTelephonePaymentRequest.getPaymentOutcome().getCode(),
+                supplemental
+        );
+
```
Comment (mrlumbu) [inline src/main/java/uk/gov/pay/api/service/telephone/CreateTelephonePaymentService.java]: Not sure of the purpose of these lines -  you are basically cloning the incoming payment outcome, but not sure why

File: src/main/java/uk/gov/pay/api/service/telephone/CreateTelephonePaymentService.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        PaymentOutcome paymentOutcome = new PaymentOutcome(
+                createTelephonePaymentRequest.getPaymentOutcome().getStatus(),
+                createTelephonePaymentRequest.getPaymentOutcome().getCode(),
+                supplemental
+        );
+        
+        State state = new State(
+                "Success",
+                true,
+                "Hello, world!",
```
Comment (mrlumbu) [inline src/main/java/uk/gov/pay/api/service/telephone/CreateTelephonePaymentService.java]: maybe make this a bit less informal

File: src/main/java/uk/gov/pay/api/service/telephone/CreateTelephonePaymentService.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+                createTelephonePaymentRequest.getEmailAddress(),
+                createTelephonePaymentRequest.getCardExpiry(),
+                createTelephonePaymentRequest.getLastFourDigits(),
+                createTelephonePaymentRequest.getFirstSixDigits(),
+                createTelephonePaymentRequest.getTelephoneNumber(),
+                "DUMMY API - RECORD NOT PERSISTED",
+                state
+        );
+        
+        return telephonePaymentResponse;
```
Comment (mrlumbu) [inline src/main/java/uk/gov/pay/api/service/telephone/CreateTelephonePaymentService.java]: you can just return in the line above

File: src/test/java/uk/gov/pay/api/resources/telephone/PaymentOutcomeValidationResourceTest.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        map.put("card_expiry", "01/08");
+        map.put("last_four_digits", "1234");
+        map.put("first_six_digits", "123456");
+
+        String payload = toJson(map);
+
+        Response response = sendPayload(payload);
+        assertThat(response.getStatus(), is(422));
+    }
+}
```
Comment (mrlumbu) [inline src/test/java/uk/gov/pay/api/resources/telephone/PaymentOutcomeValidationResourceTest.java]: comprehensive tests 👍

File: src/main/java/uk/gov/pay/api/model/telephone/State.java (hunk 1 above)
Comment (nimalank7) [reply src/main/java/uk/gov/pay/api/model/telephone/State.java]: Done

File: src/main/java/uk/gov/pay/api/model/telephone/TelephonePaymentResponse.java (hunk 2 above)
Comment (nimalank7) [reply src/main/java/uk/gov/pay/api/model/telephone/TelephonePaymentResponse.java]: Done

File: src/main/java/uk/gov/pay/api/model/telephone/TelephonePaymentResponse.java (hunk 3 above)
Comment (nimalank7) [reply src/main/java/uk/gov/pay/api/model/telephone/TelephonePaymentResponse.java]: Done

File: src/main/java/uk/gov/pay/api/service/telephone/CreateTelephonePaymentService.java (hunk 5 above)
Comment (nimalank7) [reply src/main/java/uk/gov/pay/api/service/telephone/CreateTelephonePaymentService.java]: Done

File: src/main/java/uk/gov/pay/api/service/telephone/CreateTelephonePaymentService.java (hunk 6 above)
Comment (nimalank7) [reply src/main/java/uk/gov/pay/api/service/telephone/CreateTelephonePaymentService.java]: Done

File: src/main/java/uk/gov/pay/api/service/telephone/CreateTelephonePaymentService.java (hunk 4 above)
Comment (nimalank7) [reply src/main/java/uk/gov/pay/api/service/telephone/CreateTelephonePaymentService.java]: I was playing around with Optionals when checking whether the JSON returned as expected. Then I realised that these lines were unnecessary so I've removed them.

File: src/main/java/uk/gov/pay/api/model/telephone/TelephonePaymentResponse.java (hunk 2 above)
Comment (mrlumbu) [reply src/main/java/uk/gov/pay/api/model/telephone/TelephonePaymentResponse.java]: Not sure if you've pushed these changes?

Comment (nimalank7) [reply src/main/java/uk/gov/pay/api/model/telephone/TelephonePaymentResponse.java]: Apologies - I forgot to do this class it's done now

**Pull request merged**